projects/observability/src/shared/components/bar-gauge/bar-gauge.component.ts

```diff
@@ -0,0 +1,102 @@
+import { ChangeDetectionStrategy, Component, ElementRef, Input, OnChanges, ViewChild } from '@angular/core';
+import { getPercentage, TypedSimpleChanges } from '@hypertrace/common';
+
+@Component({
+  selector: 'ht-bar-gauge',
+  styleUrls: ['./bar-gauge.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  template: `
+    <div class="bar-gauge">
+      <div *ngIf="this.title" class="title">{{ this.title | htDisplayTitle }}</div>
+      <div class="count">
+        {{ this.totalValue | number }} / {{ this.maxValue | number }}
+        <span class="units" *ngIf="this.units">{{ this.units }}</span>
+      </div>
+      <div class="bar">
+        <div #maxValueBar class="max-value-bar" [ngClass]="{ 'over-max-value': this.overMaxValue }">
+          <div #segmentBars class="segment-bars">
+            <div
+              *ngFor="let segment of this.barSegments"
+              class="segment-bar"
+              [ngClass]="{ 'hide-divider': this.nearMaxValue }"
+              [style.background]="segment.color"
+              [style.width.%]="segment.percentage"
+            >
+              <div class="divider"></div>
+            </div>
+          </div>
+        </div>
+      </div>
+    </div>
+  `
+})
+export class BarGaugeComponent implements OnChanges {
+  @ViewChild('maxValueBar', { read: ElementRef })
+  public maxValueBar!: ElementRef;
+
+  @ViewChild('segmentBars', { read: ElementRef })
+  public segmentBars!: ElementRef;
+
+  @Input()
+  public title?: string;
+
+  @Input()
+  public units?: string;
+
+  @Input()
+  public maxValue?: number;
+
+  @Input()
+  public segments?: Segment[] = [];
+
+  public barSegments: BarSegment[] = [];
+  public totalValue: number = 0;
+  public overMaxValue: boolean = false;
+  public nearMaxValue: boolean = false;
+
+  public ngOnChanges(changes: TypedSimpleChanges<this>): void {
+    if (changes.segments || changes.maxValue) {
+      this.totalValue = this.calcTotalValueFromSegments(this.segments ?? []);
+      this.barSegments = this.toBarSegments(this.segments ?? [], this.maxValue ?? 0);
+
+      this.overMaxValue = this.maxValue !== undefined && this.totalValue > this.maxValue;
+
+      this.checkNearMaxValue();
+    }
+  }
+
+  private checkNearMaxValue(): void {
+    setTimeout(() => {
+      /*
+       * We want to remove the white divider if we fill up the bar. Compare segments to
+       * max bar width minus 2px for the divider width.
+       */
+      this.nearMaxValue = this.segmentBars.nativeElement.offsetWidth >= this.maxValueBar.nativeElement.offsetWidth - 2;
```

Comment:
This feels like a very confusing way to do this, but I'm not exactly sure what we're trying to do. Are we adding dividers that are adding a fixed width to the bar? Could we measure the segments to be like ` calc( percentage - dividerWidth )` ? Any time we start measuring, or setting timeouts (let alone both) my alarm bells start going off.

Reply:
Yeah, it could be confusing to look at in code without playing with the component a bit. I'll add a clearer comment and see if that makes more sense.

Reply:
It just seems like, given the fact that we know the divider sizes, we don't need to do it dynamically like this - we could size the segments to leave room for dividers, or size the container to subtract the total divider width, right?

Reply:
Maybe I'm not following... or maybe not explaining correctly. Probably both.

Each segment is capped with the white divider line to create a visual barrier between the different segments (or the rightmost segment and the gray background). But the point of this code is to remove that white line for the right most bar when it is close to maxing out and touching the end of the gray background bar. We want this so that the white divider doesn't break the visual edge of the bar's perimeter due to it being white on a white background. 

<img width="440" alt="Screen Shot 2020-12-21 at 9 31 43 AM" src="https://user-images.githubusercontent.com/45181984/102805639-6122c980-4370-11eb-8309-7fb1b49d82af.png">

Reply:
So I understood the general goal, but one thing that did confuse me is I thought all dividers were behind hid (since we're applying that to each divider) - but it looks like we then use css to ignore the non-final ones - keeping that class only on the last segment would help clarify that confusion. 

My general ask I think holds though - it seems like there are ways to do this without having to make the whole component async + layout sensitive. A few different ideas (may not all work):
- do the comparison based on the values and not the measured pixels (can do this part of the segment-building logic), removing the bar if the data is at 100%. There may be a couple of gauge size/data combinations that will slightly misbehave, but they'd be exceedingly rare (and would only result in a 1px visual issue).
- Make the divider transparent rather than white (may need to change the div layout, don't remember how the background colors were applied)
- Make the container clip any overflow

Reply:
Tried a few things. Managed to get rid of the setTimeout, but didn't like any other solutions.

Reply:
I'm satisfied enough without setTimeout, but I maintain there's a css-only way of accomplishing this 😉

Reply:
There may be, let's come back to it if needed. Thanks! 👍